lib/service/performance.go

```
@@ -74,7 +74,12 @@ func (s *PerformanceService) RegisterPerformance(newPerf *asset.NewPerformance,
 	if _, ok := task.Outputs[newPerf.ComputeTaskOutputIdentifier]; !ok {
 		return nil, errors.NewMissingTaskOutput(task.Key, newPerf.ComputeTaskOutputIdentifier)
 	}
-	algoOutput, ok := task.Algo.Outputs[newPerf.ComputeTaskOutputIdentifier]
+	algoTask, err := s.GetAlgoService().GetAlgo(task.AlgoKey)
```

Here you retrieve the same Algo twice, L69 and L77, you could retrieve it once L69, validate that the Algo from L69 have the same key as `task.AlgoKey` then, reuse the Algo from L69 for all the other checks.

---

So we can be sure this is the same algo that is represented by task.AlgoKey and newPerf.MetricKey? If so I will also adapt the tests.

---

It should be the same, yes if that's not the case you should return an error.